NET-414: Refactor broker's ConfigWizard

bearni95 opened the pull request:
Functionality changes:
- default plugin is `publishHttp` instead of `legacyPublishHttp`
- using [updated broker defaults](https://github.com/streamr-dev/network-monorepo/commit/dbc86d2b7df34ebdc30d55f31df45dce702e81c5)
___
**Important note for Windows**
Windows compatibility at some point. All path separators should use https://nodejs.org/api/path.html#path_path_sep e.g. via https://nodejs.org/api/path.html#path_path_join_paths
___ 

Abandon the class-driven implementation for the ConfigWizard:

```typescript
export class ConfigWizard{
    async generateOrImportPrivateKey(): Promise<Wallet>
    async selectPlugins(): Promise<Array<string>>
}
```

In favor of the suggested approach by `inquirer`'s author, much more flexible for testing and abstracting the user answers from the logic and config generation:

```typescript
let prompts: Array<inquirer.Question | inquirer.ListQuestion | inquirer.CheckboxQuestion> = [
    {
        type: 'list',
        name:'generateOrImportEthereumPrivateKey',
        message: 'Do you want to generate a new Ethereum private key or import an existing one?',
        choices: [PRIVATE_KEY_SOURCE_GENERATE, PRIVATE_KEY_SOURCE_IMPORT]
    },
    {
        type: 'input',
        name:'importPrivateKey',
        message: 'Please provide the private key to import',
        when: (answers: inquirer.Answers) => {
            return answers.generateOrImportEthereumPrivateKey === PRIVATE_KEY_SOURCE_IMPORT
        },
        validate: (input: string): string | boolean => {
            try {
                new Wallet(input)
                return true
            } catch (e) {
                if (e.message.includes('invalid hexlify value')){
                    return `Invalid privateKey provided for import: ${input}`
                } else {
                    return e.message
                }
            }
        }
    }
]

const pluginSelectorPrompt = {...}
const pluginPrompts = [{...}, {...}]

prompts = prompts.concat(pluginSelectorPrompt).concat(pluginPrompts)

const answers = await inquirer.prompt(prompts)
const config = getConfigFromAnswers(answers)
```

bearni95 committed:
feat: overwrite confirmation if the existing configuration exists; display the private key and encourage to store it after 'generate'; display the associated privateKey and address after both import and generate steps

bearni95 committed:
fix: re-implemented wizard flow using native prompt object structure from inquirer. no clue how to proceed on the tests, tho

bearni95 committed:
fix: reverted to approved, method-driven functionality

bearni95 committed:
feat: refactored ConfigWizard, tests still pending

bearni95 committed:
fix: separation of answers and config to ease further testing

bearni95 committed:
fix: validation tests completed

bearni95 committed:
fix: separated logic from validation regarding answers and config

bearni95 committed:
fix: cleared variable names and export, improved NaN error reporting on port capture

bearni95 committed:
test: complete tests for the refactor, minor code improvements; code cleanup

bearni95 committed:
fix: implemented chalk lib for terminal colors; refactored logger to avoid using mouhtfulls like logger.log

bearni95 committed:
fix: removed redundand isNaN from port validation fn

bearni95 committed:
fix: split validation and logic for storage; added relevant tests

bearni95 committed:
fix: solved merge conflicts

bearni95 committed:
fix: pre-merge

bearni95 committed:
fix: imported default ports from config.schema for each plugin; implemented Partial<Config> to allow for redundant values to be omited

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -1,38 +1,46 @@
+}
 
-import * as os from 'os'
+logger.info = (...args: any[]) => {
```
Maybe all logger methods could be functions of a logger object? If there is no need that this "plain" logger is an exceptional case.

bearni95 replied:
Maybe renaming the previous approach is easier?

```typescript
const logger = {
    print: (...args: any[]) => {
        console.log(chalk.bgWhite.black(':'), ...args)
    },
    info: (...args: any[]) => {
        console.log(chalk.bgWhite.black(':', ...args))
    },
    alert: (...args: any[]) => {
        console.log(chalk.bgYellow.black('!', ...args))
    },
    warn: (...args: any[]) => {
        console.log(chalk.bgYellow.black('!'), ...args)
    },
    error: (...args: any[]) => {
        console.log(chalk.bgRed.black('!'), ...args)
    }
}
```

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -1,38 +1,46 @@
-    generateSessionId: false,
+import * as WebsocketConfigSchema from './plugins/websocket/config.schema.json'
+import * as MqttConfigSchema from './plugins/mqtt/config.schema.json'
+//import * as HttpConfigSchema from './plugins/publishHttp/config.schema.json'
```
Remove comment?

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -1,38 +1,46 @@
+
+const DEFAULT_WS_PORT = WebsocketConfigSchema.properties.port.default
+const DEFAULT_MQTT_PORT = MqttConfigSchema.properties.port.default
+const DEFAULT_HTTP_PORT = 8585//HttpConfigSchema.properties.port.default
```
Why this is not read from configs? And the value is different from the Broker's default `httpServer` port?

bearni95 replied:
this is the one we were commenting on, the path is not the indicated since the plugin shouldn't have a default value for it.

teogeb replied:
But currently it is used in `publishHttp: { port: DEFAULT_HTTP_PORT }`. That http port question should be written into the root level httpServer block if user wants to change the default. And if user doesn't want to change, we shouldn't write any port value

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
-                message: "'Please provide the private key to import'",
-            })
-
+export let prompts: Array<inquirer.Question | inquirer.ListQuestion | inquirer.CheckboxQuestion> = [
```
Use `const` instead of `let` (there can be a temporary non-exported value to evaluate the prompts value, but export it as a const)

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
+        name:'importPrivateKey',
+        message: 'Please provide the private key to import',
+        when: (answers: inquirer.Answers) => {
+            return answers.generateOrImportEthereumPrivateKey === 'Import'
```
'Generate' and 'Import' could be constants are we are referring to the value in multiple places

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
+prompts = prompts.concat(pluginSelectorPrompt).concat(pluginPrompts)
+
+export const getConfigFromAnswers = (answers: any): Config => {
+    const config = DEFAULT_CONFIG
```
Should we clone the DEFAULT_CONFIGS to the config variable? Now we are modifying the exported constant value (which is used e.g. in tests)

bearni95 replied:
Something like `const config = Object.assign({}, DEFAULT_CONFIG)`?

teogeb replied:
Or https://lodash.com/docs/4.17.15#cloneDeep, if you change e.g. the contents of config.plugins

bearni95 replied:
is it worth installing a whole new package just for this functionality?

teogeb replied:
True, maybe easiest solution is to clone both config and plugins:
```
const config = { ... DEFAULT_CONFIG, plugins: { ... DEFAULT_CONFIG.plugins } }
```

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
+    validate: (input: string, answers: inquirer.Answers = {}): string | boolean => {
+        try {
+            const filePath = input || answers.selectDestinationPath
+            const parentDirPath = filePath.substring(0, filePath.lastIndexOf('/'))
```
Could use https://nodejs.org/api/path.html#path_path_dirname_path

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
+    type: 'input',
+    name: 'selectDestinationPath',
+    message: `Select a path to store the generated config in `,
+    default: path.join(os.homedir(), '.streamr/broker-config.json'),
```
Not related to this PR, but worth checking the Windows compatibility at some point. All path separators should use https://nodejs.org/api/path.html#path_path_sep e.g. via https://nodejs.org/api/path.html#path_path_join_paths

bearni95 replied:
Added the note to the PR description, will make a better explanation out of it since Windows support is a whole thing

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
+            answers.parentDirExists = existsSync(parentDirPath)
+            answers.fileExists = existsSync(filePath)
+
+            if (!answers.parentDirExists && !answers.fileExists){
```
Minor enhancement: just return `(!answers.parentDirExists && !answers.fileExists)` without the if statement

bearni95 replied:
Actually `(answers.parentDirExists || answers.fileExists)` but yup, good improvement

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
-                {name:'HttpPublish'}
-            ]
-        })
+const pluginTemplates = {
```
This is a constant -> rename to PLUGIN_TEMPLATES?

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,202 @@
-        tmpDataDir = mkdtempSync(path.join(os.tmpdir(), 'broker-test-config-wizard'))
-    })
+    const importPrivateKeyPrompt = prompts[1]
+    const validatePortPrompt = prompts[3]
```
Rename to `portPrompt`?

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,202 @@
+        it ('happy path, prefixed', () => {
+            const validate = importPrivateKeyPrompt.validate!
+            const privateKey = Wallet.createRandom().privateKey
+            expect(privateKey.length).toBe(66)
```
No need to assert the internal variable (also at line 23)

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
 
-            if (!overwrite) {
-                return this.storeConfig()
+            answers.parentDirPath = parentDirPath
```
Do we need to store parentDirPath separately? If we already store the selectDestinationPath

bearni95 replied:
We do in order to create, if necessary, the folder on the later logic step

teogeb replied:
Yes, we can calculate it there easily. If I understood correctly that selectDestinationPath is always path.dirname(selectDestinationPath), e.g. `/foo/bar/file.json` -> `/foo/bar/`

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
-            if (!overwrite) {
-                return this.storeConfig()
+            answers.parentDirPath = parentDirPath
+            answers.parentDirExists = existsSync(parentDirPath)
```
Also are these two variables needed? The facts about file existence are not answers that the user provided

bearni95 replied:
Once again, needed later to generate the `~/.streamr` path, by default, if it didn't exist

teogeb replied:
But less state to maintain if we calculate it when we need it

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,202 @@
 
-        const importedWallet: Wallet = await wizard.generateOrImportPrivateKey()
-        expect(importedWallet.privateKey).toEqual(privateKey)
+        it ('happy path with overwrite destination', () => {
```
The tested functionality was bit unclear (if the similar test at line 101 does test the overwrite functionality then maybe this could be removed)

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,164 @@ const DefaultConfig: Config = {
-                    payloadMetadata: false
-                }
+const pluginPrompts = Object.values(pluginTemplates).map((plugin, i) => {
+    const pluginName = Object.keys(pluginTemplates)[i]
```
We can't know the iteration order of the object keys

bearni95 replied:
Can't we?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values

the `for...in` loop for `Object.values` should do the same string sorting as the one provided in `Object.keys`, right?

teogeb replied:
You are right 👍 Wasn't aware of that the matching ordering is a documented feature. 

(Still it would be maybe more readable and more straightforward if it would be something like)
```
Object.keys(pluginTemplates).map((pluginName) => {
  const plugin = pluginTemplates[name]
})
```

bearni95 committed:
fix: round of comments

- rephrased the logger object for clarity
- removed comments from imports on http ports
- exporting consts instead of lets
- used constants for Generate and Import strings
- using path.dirname to extract dirnames
- using object clone when interacting with DEFAULT_CONFIG
- simplified boolean return on path validation prompt
- renamed PLUGIN_TEMPLATES
- removed redundant tests

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,202 @@
+            }
+            const config = getConfigFromAnswers(answers)
+            expect(config.ethereumPrivateKey).toBeDefined()
+            expect(config.ethereumPrivateKey.match(/^(0x)?[a-f0-9]{64}$/)).not.toBe(null)
```
Could use Jest's `stringMatching` (https://jestjs.io/docs/expect#expectstringmatchingstring--regexp) for all regexp checks

bearni95 committed:
fix: using jest expect.toMatch for regexp matches; reorganized impoorts and consts

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +87,168 @@ const DefaultConfig: Config = {
+        mkdirSync(answers.parentDirPath)
     }
+   
+    writeFileSync(answers.selectDestinationPath, JSON.stringify(config))
```
Should write the JSON with indentations (`JSON.stringify(this.config, null, 2)`), like it was previously

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +87,168 @@ const DefaultConfig: Config = {
+const PLUGIN_TEMPLATES = {
+    websocket: { port: DEFAULT_WS_PORT },
+    mqtt: { port: DEFAULT_MQTT_PORT },
+    publishHttp: { port: DEFAULT_HTTP_PORT }
```
The `publishHttp` plugin doesn't have a `port` property in the plugin configuration. It uses whatever port is configured to `httpServer`. So we should maybe set the value to httpServer.port (line 62) if user wants to override the port. 

And if user doesn't override some port (a plugin port or `httpServer` port) it would be maybe easiest to leave the configuration out. For HTTP server we can leave the whole `httpServer` block out in that case.

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -57,7 +63,6 @@ const DefaultConfig: Config = {
         privateKeyFileName: null,
```
These `null` values are not needed as null is the default

bearni95 replied:
since the `httpServer` property has it's own type the values are required to be present, even if null

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +87,168 @@ const DefaultConfig: Config = {
-            logger.warn(`Non-numeric value [${valueString}] provided`)
-            return await this.promptNumberWithDefault(prompt, defaultValue)
-        }
+const promptValues = {
```
Would maybe give some more specific name? Or just two constants, e.g. `PRIVATE_KEY_SOURCE_GENERATE` and `PRIVATE_KEY_SOURCE_IMPORT`

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,189 @@
+            config = DEFAULT_CONFIG as Config
+        })
+
+        it ('happy path with overwrite destination', () => {
```
This test used to be a test the actual storing (`expect(existsSync(configFileLocation)).toBe(true)`) and now it tests the validation logic. Could move into "storagePrompt validation" section if the main focus is on validation.

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +87,168 @@ const DefaultConfig: Config = {
+    const config = Object.assign({}, DEFAULT_CONFIG)
+    const pluginNames = Object.keys(PLUGIN_TEMPLATES)
+    const PLUGIN_TEMPLATESArray = Object.values(PLUGIN_TEMPLATES)
+    for (let i = 0; i < PLUGIN_TEMPLATESArray.length; i++){
```
This loop could be simplified? Something like this:
```
pluginNames.forEach((name) => {
  const template = PLUGIN_TEMPLATES[name]
  ...
})
```

bearni95 committed:
fix: comments corrected

- improved cloning of base options object
- improved PR description
- simplified iterators involving PLUGIN_TEMPLATES
- added JSON indentation on file write
- reworked HTTP logic
- using static consts for "generate" and "import" presentation
- moved validation test back into it's block

bearni95 committed:
fix: made DEFAULT_CONFIG type any instead of Partial<Config> to avoid pre-felling unnecessary fields

bearni95 committed:
fix: improved http logic

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +84,170 @@ const DefaultConfig: Config = {
+        const template = PLUGIN_TEMPLATES[pluginName]
+        if (answers.selectPlugins && answers.selectPlugins.includes(pluginName)){
+            // the publishHttp plugin is special, it needs to be added to the config after the other plugins
+            if (pluginName === 'publishHttp') {
```
If we refer to this value twice, it should be a constant (maybe all plugin names could be)

bearni95 replied:
with the latest commit it is not repeated anymore, seems quite redundant to implement all plugin names as constants for single usage

teogeb replied:
It is still at line 126 and line 173
```
publishHttp: { port: DEFAULT_HTTP_PORT }
if (pluginName === 'publishHttp') {
```

bearni95 committed:
fix: added behaviour for empty plugin ports when default selected; added relevant tests

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -164,12 +171,17 @@ export const getConfigFromAnswers = (answers: any): Config => {
                 }
             } else {
+                // use provided a custom value, fill in
                 config.plugins![pluginName] = { ...template, port: answers[`${pluginName}Port`] }
```
Is the `...template` part needed anymore?

bearni95 replied:
if there ever were fields to be added into the template that should be carried along to the generated config the current structure would allow them to

teogeb replied:
Ok 👍 Do we support that future feature for `publishHttp` too?

bearni95 replied:
not at the moment since it has it's own special path having to do with `httpServer`

teogeb replied:
This may be confusing if someone assumes that the feature is working ok for all plugins. Maybe easiest is to drop that redundant feature

bearni95 replied:
Applied proposed solution

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -164,12 +171,17 @@ export const getConfigFromAnswers = (answers: any): Config => {
-            if (pluginName === 'publishHttp') {
+            if (answers[`${pluginName}Port`] === template.port){
+                // user selected default value, enable without specifying the port
+                config.plugins![pluginName] = {}
```
If we modify `config.plugins![pluginName]` in each section, maybe that could be outside the `if`s

E.g.
```
let pluginConfig = {}
if ... {
...
}
config.plugins![pluginName] = pluginConfig
```

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -52,12 +67,6 @@ const DefaultConfig: Config = {
-    apiAuthentication: null,
     plugins: {
         legacyWebsocket: {
             port: DEFAULT_LEGACY_WS_PORT
```
This could be removed

bearni95 replied:
They've been, hence the red coloring

teogeb replied:
Mean the `DEFAULT_LEGACY_WS_PORT`

bearni95 replied:
👍

bearni95 committed:
fix: simplified if statement for getConfigFromAnswers

bearni95 committed:
fix: removed ws legacy plugin from default object

bearni95 committed:
fix: removed redundant template conditional assignation

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +88,175 @@ const DefaultConfig: Config = {
+export const getConfigFromAnswers = (answers: any): Config => {
+    const config = { ... DEFAULT_CONFIG, plugins: { ... DEFAULT_CONFIG.plugins } }
+
+    const pluginNames = Object.keys(PLUGIN_TEMPLATES)
```
Now this is not a template anymore, rename to PLUGIN_DEFAULT_PORTS or something? E.g.

```
{
  websocket: DEFAULT_WS_PORT
}
```

bearni95 committed:
fix: renamed PLUGIN_TEMPLATES into DEFAULT_PLUGIN_PORTS; modified the logic to accomodate it

bearni95 committed:
fix: plugin names are consts now

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -82,200 +88,181 @@ const DefaultConfig: Config = {
-                {name:'HttpPublish'}
-            ]
-        })
+    return config as Config
```
The JSON is not a `Config` type (which is the type definition for the configuration file that we get **after** we parse it by `Ajv`)

bearni95 committed:
fix: removed Config types; improved typing in answers object from any to inquirer.Answers

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,178 +1,207 @@
+        })
+
+        beforeEach(() => {
+            config = { ... DEFAULT_CONFIG, plugins: { ... DEFAULT_CONFIG.plugins } }
```
Why is this needed? The contents of the config is not relevant for these tests, could use just a constant `MOCK_CONFIG = {}`

bearni95 replied:
A `plugins` key would be needed to satisfy the logic. Can we revert to  `config = DEFAULT_CONFIG` ?

teogeb replied:
But the value is assigned only in beforeEach, so it is a constant (and if you don't change the contents in any way, why to make a copy)

bearni95 replied:
So can we revert to `config = DEFAULT_CONFIG`?

teogeb replied:
`createStorageFile` is the method which we are testing here. It accepts `config: any`. So an empty JSON is a valid input (if no need to be more complicated, the simplest input is the best)

bearni95 replied:
Terrific, implemented on the next commit!

bearni95 committed:
fix: using {} instead of a Config in createStorageFile beforeEach

bearni95 committed:
fix: using const for config inside createStorageFile test

teogeb reviewed (approved):
LGTM!

bearni95 committed:
Merge branch 'main' into NET-414
